File: sqlparser/src/sql/SqlCondition.java
//@
package cdata.sql;
//@
/*#
using RSSBus.core;
using RSSBus;
namespace CData.Sql {
#*/

public /*@*/final/*@*//*#sealed#*/ class SqlCondition extends SqlConditionNode {
  private /*@*/final/*@*/ /*#readonly#*/  SqlLogicalOperator logicOp;
  private /*@*/final/*@*/ /*#readonly#*/  SqlExpression lhs, rhs;

  public SqlCondition(SqlExpression lhs, SqlLogicalOperator op, SqlExpression rhs)  {
    this.lhs = lhs;
    this.logicOp = op;
    this.rhs = rhs;
  }

  public /*#override#*/ boolean isEvaluatable() {
    if (lhs != null && rhs != null) {
      if (rhs.isEvaluatable() && lhs.isEvaluatable()) {
        return true;
      } else if (rhs.isEvaluatable()) {
        if (SqlLogicalOperator.And == this.getLogicOp()) {
          return isFalseValue(rhs);
        } else if (SqlLogicalOperator.Or == this.getLogicOp()) {
          return isTrueValue(rhs);
        }
      } else if (lhs.isEvaluatable()) {
        if (SqlLogicalOperator.And == this.getLogicOp()) {
          return isFalseValue(lhs);
        } else if (SqlLogicalOperator.Or == this.getLogicOp()) {
          return isTrueValue(lhs);
        }
      }
    }
    return false;
  }

  public /*#override#*/ SqlValue evaluate() {
    if (isEvaluatable()) {
      SqlValue trueValue = new SqlValue(SqlValueType.BOOLEAN, "TRUE");
      SqlValue falseValue = new SqlValue(SqlValueType.BOOLEAN, "FALSE");
      if (logicOp == SqlLogicalOperator.And) {
        if ((lhs.isEvaluatable() && isFalseValue(lhs))
          || (rhs.isEvaluatable() && isFalseValue(rhs))) {
          return falseValue;
        } else {
          if ((lhs.isEvaluatable() && isTrueValue(lhs))
            && (rhs.isEvaluatable() && isTrueValue(rhs))) {
            return trueValue;
          }
        }
      } else if (logicOp == SqlLogicalOperator.Or) {
        if ((lhs.isEvaluatable() && isTrueValue(lhs))
          || (rhs.isEvaluatable() && isTrueValue(rhs))) {
          return trueValue;
        } else {
          return falseValue;
        }
      } else {
        return falseValue;
      }
    }
    return SqlValue.getNullValueInstance();
  }

  public /*#override#*/ boolean isLeaf() { return false; }

  public /*#override#*/ void prepare() throws Exception {
    if (lhs instanceof SqlConditionNode) {
      ((SqlConditionNode)lhs).prepare();
    }
    if (rhs instanceof SqlConditionNode) {
      ((SqlConditionNode)rhs).prepare();
    }
  }

  public SqlExpression getLeft() {
    return this.lhs;
  }

  public SqlExpression getRight() {
    return this.rhs;
  }

  public SqlLogicalOperator getLogicOp() {
    return logicOp;
  }

  public /*#override#*/ void accept(ISqlQueryVisitor visitor) throws Exception {
    SqlExpression l = this.getLeft();
    SqlExpression r = this.getRight();
    if (l instanceof SqlCondition) {
      ((SqlCondition) l).accept(visitor);
    } else if (l instanceof SqlConditionNode) {
      visitor.visit((SqlConditionNode) l);
    }

    if (r instanceof SqlCondition) {
      ((SqlCondition) r).accept(visitor);
    } else if (r instanceof SqlConditionNode) {
      visitor.visit((SqlConditionNode) r);
    }
  }

  public /*#override#*/ Object clone() {
    SqlExpression l = getLeft();
    SqlExpression r = getRight();
    if (l instanceof SqlConditionNode) {
      l = (SqlExpression)((SqlConditionNode)l).clone();
    }
    if (r instanceof SqlConditionNode) {
      r = (SqlExpression)((SqlConditionNode)r).clone();
    }
    return new SqlCondition(l, getLogicOp(), r);
  }

  private static boolean isFalseValue(SqlExpression expr) {
    boolean isFalse = false;
    try {
      SqlValue v = expr.evaluate();
      if (v != null) {
        if (SqlValueType.BOOLEAN == v.getValueType()) {
          isFalse = false == v.getValueAsBool(true);
        }
      }
    } catch (Exception ex) {}
    return isFalse;
  }

  private static boolean isTrueValue(SqlExpression expr) {
    boolean isTrue = false;
    try {
      SqlValue v = expr.evaluate();
      if (v != null) {
        if (SqlValueType.BOOLEAN == v.getValueType()) {
          isTrue = true == v.getValueAsBool(false);
        }
      }
    } catch (Exception ex) {}
    return isTrue;
  }
}

File: sqlparser/src/sql/csgen/SqlCondition.cs
using System;
using System.Collections;
using System.IO;
using System.Threading;
using System.Net;
using System.Net.Sockets;
using System.Text;
using RSSBus.core.j2cs;



using RSSBus.core;
using RSSBus;
namespace CData.Sql {


public sealed class SqlCondition : SqlConditionNode {
  private  readonly  SqlLogicalOperator logicOp;
  private  readonly  SqlExpression lhs, rhs;

  public SqlCondition(SqlExpression lhs, SqlLogicalOperator op, SqlExpression rhs)  {
    this.lhs = lhs;
    this.logicOp = op;
    this.rhs = rhs;
  }

  public override bool IsEvaluatable() {
    if (lhs != null && rhs != null) {
      if (rhs.IsEvaluatable() && lhs.IsEvaluatable()) {
        return true;
      } else if (rhs.IsEvaluatable()) {
        if (SqlLogicalOperator.And == this.GetLogicOp()) {
          return IsFalseValue(rhs);
        } else if (SqlLogicalOperator.Or == this.GetLogicOp()) {
          return IsTrueValue(rhs);
        }
      } else if (lhs.IsEvaluatable()) {
        if (SqlLogicalOperator.And == this.GetLogicOp()) {
          return IsFalseValue(lhs);
        } else if (SqlLogicalOperator.Or == this.GetLogicOp()) {
          return IsTrueValue(lhs);
        }
      }
    }
    return false;
  }

  public override SqlValue Evaluate() {
    if (IsEvaluatable()) {
      SqlValue trueValue = new SqlValue(SqlValueType.BOOLEAN, "TRUE");
      SqlValue falseValue = new SqlValue(SqlValueType.BOOLEAN, "FALSE");
      if (logicOp == SqlLogicalOperator.And) {
        if ((lhs.IsEvaluatable() && IsFalseValue(lhs))
          || (rhs.IsEvaluatable() && IsFalseValue(rhs))) {
          return falseValue;
        } else {
          if ((lhs.IsEvaluatable() && IsTrueValue(lhs))
            && (rhs.IsEvaluatable() && IsTrueValue(rhs))) {
            return trueValue;
          }
        }
      } else if (logicOp == SqlLogicalOperator.Or) {
        if ((lhs.IsEvaluatable() && IsTrueValue(lhs))
          || (rhs.IsEvaluatable() && IsTrueValue(rhs))) {
          return trueValue;
        } else {
          return falseValue;
        }
      } else {
        return falseValue;
      }
    }
    return SqlValue.GetNullValueInstance();
  }

  public override bool IsLeaf() { return false; }

  public override void Prepare() {
    if (lhs is SqlConditionNode) {
      ((SqlConditionNode)lhs).Prepare();
    }
    if (rhs is SqlConditionNode) {
      ((SqlConditionNode)rhs).Prepare();
    }
  }

  public SqlExpression GetLeft() {
    return this.lhs;
  }

  public SqlExpression GetRight() {
    return this.rhs;
  }

  public SqlLogicalOperator GetLogicOp() {
    return logicOp;
  }

  public override void Accept(ISqlQueryVisitor visitor) {
    SqlExpression l = this.GetLeft();
    SqlExpression r = this.GetRight();
    if (l is SqlCondition) {
      ((SqlCondition) l).Accept(visitor);
    } else if (l is SqlConditionNode) {
      visitor.Visit((SqlConditionNode) l);
    }

    if (r is SqlCondition) {
      ((SqlCondition) r).Accept(visitor);
    } else if (r is SqlConditionNode) {
      visitor.Visit((SqlConditionNode) r);
    }
  }

  public override Object Clone() {
    SqlExpression l = GetLeft();
    SqlExpression r = GetRight();
    if (l is SqlConditionNode) {
      l = (SqlExpression)((SqlConditionNode)l).Clone();
    }
    if (r is SqlConditionNode) {
      r = (SqlExpression)((SqlConditionNode)r).Clone();
    }
    return new SqlCondition(l, GetLogicOp(), r);
  }

  private static bool IsFalseValue(SqlExpression expr) {
    bool isFalse = false;
    try {
      SqlValue v = expr.Evaluate();
      if (v != null) {
        if (SqlValueType.BOOLEAN == v.GetValueType()) {
          isFalse = false == v.GetValueAsBool(true);
        }
      }
    } catch (Exception ex) {}
    return isFalse;
  }

  private static bool IsTrueValue(SqlExpression expr) {
    bool isTrue = false;
    try {
      SqlValue v = expr.Evaluate();
      if (v != null) {
        if (SqlValueType.BOOLEAN == v.GetValueType()) {
          isTrue = true == v.GetValueAsBool(false);
        }
      }
    } catch (Exception ex) {}
    return isTrue;
  }
}
}

